rancher/rancher
# Add MachineProvisionImagePullPolicy setting

# This will not be merged. Waiting for  release v2.8! 
I just opened the PR now to get reviews / opinions on the code.

## Issue:  #42294 

 
## Problem

A client needed have an air-gapped environment and was having problems wile creating a new cluster. That happens due to the `ImagePullPolicy`  of the `...-machine-provisioning` job being set to `Always`.

## Solution

As discussed on the jira ticket the idea is to enable the client to set the image pull policy through global settings. 
If this is not configured the behavior will remain the same, that's why the default value is Always. 
The setting will be enabled in initialization (once set it cannot be changed unless rancher is restarted).

## Some questions: 

- [x] Are the function / variable names ok ? 
- [x] Should GetMachineProvisionImagePullPolicy() exist? I did it to avoid errors if the user set an invalid ImagePullPolicy.  
- [x] GetMachineProvisionImagePullPolicy() Should be on settings.go? I left it there because there are other "getters" on the file. 
- [x] If the user set the ImagePullPolicy to Never without the image the pod will get stuck in ErrImageNeverPull, do we allow "Never" as a valid input? 
 
## Testing


## Engineering Testing
### Manual Testing


Running rancher locally, the process was basically:
*  set the  env `CATTLE_MACHINE_PROVISION_IMAGE_PULL_POLICY`
*  start rancher/rancher
*  create a downstream cluster using the UI.
*  `kubectl edit job {JOB-NAME} -n fleet-default` and check the ImagePullPolicy.
*  `kubectl describe pod {POD-NAME} -n fleet-default` and check if the image was pulled or respected the value set. 

I did that test with `CATTLE_MACHINE_PROVISION_IMAGE_PULL_POLICY`  set to:  `""` (empty), `Always`, `Never`, `IfNotPresent`, `wrong` and without it set. 




### Automated Testing

* Test types added/modified:
    * Unit
    * None
    If the GetMachineProvisionImagePullPolicy is Ok i'll do unit test for it. Just waiting to confirm if it is ok. 

## QA Testing Considerations

 
### Regressions Considerations
